### PR TITLE
fix(profiling): use a pool of `Sample` to reduce memory pressure

### PR DESCRIPTION
**Pull request by taegyunkim**
## Checklist

- [ ] Change(s) are motivated and described in the PR description
- [ ] Testing strategy is described if automated tests are not included in the PR
- [ ] Risks are described (performance impact, potential for breakage, maintainability)
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [ ] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [ ] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [ ] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Description motivates each change
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Change is maintainable (easy to change, telemetry, documentation)
- [ ] Release note makes sense to a user of the library
- [ ] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp [hunk 1]
```diff
@@ -30,18 +30,36 @@ Datadog::SampleManager::set_timeline(bool enable)
...
 
 void
 Datadog::SampleManager::drop_sample(Datadog::Sample* sample)
 {
-    delete sample; // NOLINT(cppcoreguidelines-owning-memory)
+    // Reset the sample's state
+    sample->clear_buffers();
+
+    {
+        const std::lock_guard<std::mutex> lock(pool_mutex);
```
[the remark this answers is not in the document]
Comment (danielsn) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp:53]: https://www.boost.org/doc/libs/1_60_0/doc/html/boost/lockfree/stack.html

**Comment by taegyunkim** [comment]
* **#9617** <a href="https://app.graphite.dev/github/pr/DataDog/dd-trace-py/9617?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @taegyunkim and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp [hunk 2]
```diff
@@ -30,18 +30,36 @@ Datadog::SampleManager::set_timeline(bool enable)
...
+    // Create a new sample if the pool is empty
     return new Datadog::Sample(type_mask, max_nframes); // NOLINT(cppcoreguidelines-owning-memory)
 }
 
 void
 Datadog::SampleManager::drop_sample(Datadog::Sample* sample)
 {
-    delete sample; // NOLINT(cppcoreguidelines-owning-memory)
+    // Reset the sample's state
+    sample->clear_buffers();
```
[the remark this answers is not in the document]
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp:48]: https://github.com/DataDog/dd-trace-py/blob/2b14b3e4f5a700bbe7de119cd0846a1cb2b4ff52/ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp#L98-L104

Leave it around, `values` are set to all zeroes, `labels` and `locations` are `clear()`ed keeping the memory.

File: ddtrace/internal/datadog/profiling/dd_wrapper/include/sample_manager.hpp [hunk 3]
```diff
@@ -18,7 +18,9 @@ class SampleManager
   private:
     static inline unsigned int max_nframes{ g_default_max_nframes };
     static inline SampleType type_mask{ SampleType::All };
-    static inline std::mutex init_mutex{};
+
+    static inline std::mutex pool_mutex{};
+    static inline std::vector<std::unique_ptr<Sample>> sample_pool{};
```
[the remark this answers is not in the document]
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/dd_wrapper/include/sample_manager.hpp]: Done

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp [hunk 4]
```diff
@@ -30,18 +30,36 @@ Datadog::SampleManager::set_timeline(bool enable)
...
+
+    {
+        const std::lock_guard<std::mutex> lock(pool_mutex);
+        // Return the sample to the pool
+        sample_pool.emplace_back(sample); // NOLINT(cppcoreguidelines-owning-memory)
```
[the remark this answers is not in the document]
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp]: I've switched to use `boost::lockfree::queue`. We can configure it to be fixed sized. 
https://www.boost.org/doc/libs/1_85_0/doc/html/boost/lockfree/fixed_sized.html

File: ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (taegyunkim) [reply ddtrace/internal/datadog/profiling/dd_wrapper/src/sample_manager.cpp]: Yeah, used queue.

**Comment by taegyunkim** [comment]
Hmm, parametric tests times out and keeps failing when I add boost. Maybe it takes too long to compile? 

https://github.com/DataDog/dd-trace-py/actions/runs/9874600235/job/27272148666?pr=9617

**Pull request closed without merging**